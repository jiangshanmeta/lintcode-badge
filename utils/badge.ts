export const contentPresets = {
    rank: {
        presetLabel: "Rank",
        badgeLabel: "Rank",
        badgeValue: "rank",
        badgeValueDisplay: "Rank",
    },
    solved: {
        presetLabel: "Solved",
        badgeLabel: "Solved",
        badgeValue: "solved",
        badgeValueDisplay: "Solved",
    },
    "solved-over-total": {
        presetLabel: "Solved over total",
        badgeLabel: "Solved",
        badgeValue: "solvedOverTotal",
        badgeValueDisplay: "Solved over total",
    },
    "solved-percentage": {
        presetLabel: "Solved percentage",
        badgeLabel: "Solved",
        badgeValue: "solvedPercentage",
        badgeValueDisplay: "Solved percentage",
    },
} as const;

export const styles = {
    plastic: { width: "99px", height: "18px" },
    flat: { width: "81px", height: "20px" },
    "flat-square": { width: "125px", height: "20px" },
    "for-the-badge": { width: "210px", height: "28px" },
    social: { width: "98px", height: "20px" },
} as const;

export type ContentPresetName = keyof typeof contentPresets;

type Style = keyof typeof styles;

export interface Badge {
  username: string;
  style: Style;
  labelColor: string;
  color: string;
  contentPreset: ContentPresetName | "custom";
  label: string;
  value: ContentPresetName;
  showLogo: boolean;
  logoColor: string;
}

export const DEFAULT_BADGE: Badge = {
    username: "jiangshanmeta",
    style: "for-the-badge",
    labelColor: "black",
    color: "#12B4FF",
    contentPreset: "solved-over-total",
    label: "Solved",
    value: "solved-over-total",
    showLogo: false,
    logoColor: "yellow",
};


export const getUrl = (badge: Badge): string =>
    `https://img.shields.io/badge/dynamic/json?style=${
        badge.style
    }&labelColor=${encodeURIComponent(
        badge.labelColor
    )}&color=${encodeURIComponent(badge.color)}&label=${encodeURIComponent(
        badge.label
    )}&query=${
        contentPresets[badge.value].badgeValue
    }&url=https%3A%2F%2Flintcode-badge.vercel.app%2Fapi%2Fusers%2F${encodeURIComponent(
        badge.username
    )}`;

export const getMarkdown = (badge: Badge): string =>
    `[![LeetCode user ${badge.username}](${getUrl(
        badge
    )})](https://lintcode.com/user/${badge.username}/)`;
