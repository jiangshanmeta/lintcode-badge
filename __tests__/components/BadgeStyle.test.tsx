import {
    fireEvent,
    render, 
    screen
} from '@testing-library/react';

import BadgeStyle from 'components/BadgeStyle'
import {DEFAULT_BADGE} from 'utils/badge'


describe('BadgeStyle Component',()=>{

    it('should render five candicates',()=>{
        const setBadge = jest.fn()

        const {
            container
        } = render(<BadgeStyle badge={DEFAULT_BADGE} setBadge={setBadge} />)

        expect(container.getElementsByClassName('MuiFormControlLabel-root').length).toBe(5)
        expect(container.innerHTML).toMatchSnapshot()
    })

    it('should call setBadge when select',()=>{
        const setBadge = jest.fn()

        render(<BadgeStyle badge={DEFAULT_BADGE} setBadge={setBadge} />)

        fireEvent.click(screen.getByAltText(/plastic/))
        
        expect(setBadge).toBeCalledTimes(1)

        let firstParam = setBadge.mock.calls[0][0]
        if(typeof firstParam === 'function'){
            firstParam = firstParam(DEFAULT_BADGE)
        }

        expect(firstParam).toEqual({
            ...DEFAULT_BADGE,
            style:'plastic'
        });
        
    })

})