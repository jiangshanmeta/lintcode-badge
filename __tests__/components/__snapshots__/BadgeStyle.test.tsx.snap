// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeStyle Component should render five candicates 1`] = `"<fieldset class="MuiFormControl-root"><legend class="MuiFormLabel-root">Style</legend><div class="MuiFormGroup-root" role="radiogroup" aria-label="style" style="margin-top: 8px;"><label class="MuiFormControlLabel-root"><span class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorPrimary MuiIconButton-colorPrimary" aria-disabled="false"><span class="MuiIconButton-label"><input class="PrivateSwitchBase-input-4" name="style" type="radio" value="plastic"><div class="PrivateRadioButtonIcon-root-5"><svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"></path></svg><svg class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"></path></svg></div></span><span class="MuiTouchRipple-root"></span></span><span class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"><img src="style-plastic.svg" alt="badge style &quot;plastic&quot;" width="99px" height="18px"></span></label><label class="MuiFormControlLabel-root"><span class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorPrimary MuiIconButton-colorPrimary" aria-disabled="false"><span class="MuiIconButton-label"><input class="PrivateSwitchBase-input-4" name="style" type="radio" value="flat"><div class="PrivateRadioButtonIcon-root-5"><svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"></path></svg><svg class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"></path></svg></div></span><span class="MuiTouchRipple-root"></span></span><span class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"><img src="style-flat.svg" alt="badge style &quot;flat&quot;" width="81px" height="20px"></span></label><label class="MuiFormControlLabel-root"><span class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorPrimary MuiIconButton-colorPrimary" aria-disabled="false"><span class="MuiIconButton-label"><input class="PrivateSwitchBase-input-4" name="style" type="radio" value="flat-square"><div class="PrivateRadioButtonIcon-root-5"><svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"></path></svg><svg class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"></path></svg></div></span><span class="MuiTouchRipple-root"></span></span><span class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"><img src="style-flat-square.svg" alt="badge style &quot;flat square&quot;" width="125px" height="20px"></span></label><label class="MuiFormControlLabel-root"><span class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorPrimary" aria-disabled="false"><span class="MuiIconButton-label"><input class="PrivateSwitchBase-input-4" name="style" type="radio" value="for-the-badge" checked=""><div class="PrivateRadioButtonIcon-root-5 PrivateRadioButtonIcon-checked-7"><svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"></path></svg><svg class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"></path></svg></div></span><span class="MuiTouchRipple-root"></span></span><span class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"><img src="style-for-the-badge.svg" alt="badge style &quot;for the-badge&quot;" width="210px" height="28px"></span></label><label class="MuiFormControlLabel-root"><span class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorPrimary MuiIconButton-colorPrimary" aria-disabled="false"><span class="MuiIconButton-label"><input class="PrivateSwitchBase-input-4" name="style" type="radio" value="social"><div class="PrivateRadioButtonIcon-root-5"><svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"></path></svg><svg class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"></path></svg></div></span><span class="MuiTouchRipple-root"></span></span><span class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"><img src="style-social.svg" alt="badge style &quot;social&quot;" width="98px" height="20px"></span></label></div></fieldset>"`;
